File: app/src/main/java/net/chris/demo/guardian/model/inject/NetworkComponent.java
package net.chris.demo.guardian.model.inject;

import net.chris.demo.guardian.BuildConfig;
import net.chris.demo.guardian.MainService;

import javax.inject.Singleton;

import dagger.Component;

@Singleton
@Component(modules = {NetworkModule.class})
public abstract class NetworkComponent {

    public abstract void inject(MainService service);

    private static NetworkComponent instance;

    public static final NetworkComponent getInstance() {
        if (instance == null) {
            synchronized (NetworkComponent.class) {
                if (instance == null) {
                    instance = DaggerNetworkComponent.builder()
                            .networkModule(new NetworkModule(BuildConfig.REQ_ENDPOINT, BuildConfig.DEBUG))
                            .build();
                }
            }
        }
        return instance;
    }

    public static final void uninject() {
        instance = null;
    }

}

File: app/src/main/kotlin/net/chris/demo/guardian/model/inject/NetworkComponent.kt
package net.chris.demo.guardian.model.inject

import dagger.Component
import net.chris.demo.guardian.BuildConfig
import net.chris.demo.guardian.MainService
import javax.inject.Singleton

@Singleton
@Component(modules = arrayOf(NetworkModule::class))
abstract class NetworkComponent {

    abstract fun inject(service: MainService)

    companion object {

        internal var instance: NetworkComponent? = DaggerNetworkComponent.builder()
                .networkModule(NetworkModule(BuildConfig.REQ_ENDPOINT, BuildConfig.DEBUG))
                .build()

        private fun getInstance(): NetworkComponent? {
            if (instance == null) {
                synchronized(NetworkComponent::class.java) {
                    if (instance == null) {
                        instance = DaggerNetworkComponent.builder()
                                .networkModule(NetworkModule(BuildConfig.REQ_ENDPOINT, BuildConfig.DEBUG))
                                .build()
                    }
                }
            }
            return instance
        }

        fun inject(service: MainService) = getInstance()?.inject(service)

        fun unInject() {
            instance = null
        }
    }

}
